modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/metal/js/components/Calculator/Calculator.es.js

```diff
@@ -0,0 +1,231 @@
+import templates from './Calculator.soy.js';
+
+/**
+ * Calculator.
+ * @extends Component
+ */
+
+class Calculator extends Component {
+
+	static STATE = {
+		calculatorOptions: Config.arrayOf(
+			Config.shapeOf(
+				{
+					label: Config.string(),
+					tooltip: Config.string(),
+					value: Config.string()
+				}
+			)
+		).value([]),
+
+		disableCalculatorField: Config.bool().value(false),
+
+		expression: Config.string().value(''),
+
+		expressionArray: Config.array().value([]),
+
+		index: Config.number().value(0),
+
+		resultSelected: Config.string(),
+
+		options: Config.arrayOf(
+			Config.shapeOf(
+				{
+					dataType: Config.string(),
+					name: Config.string(),
+					options: Config.arrayOf(
+						Config.shapeOf(
+							{
+								label: Config.string(),
+								name: Config.string(),
+								value: Config.string()
+							}
+						)
+					),
+					type: Config.string(),
+					value: Config.string()
+				}
+			)
+		).value([]),
+
+		optionsRepeatable: Config.arrayOf(
+			Config.shapeOf(
+				{
+					dataType: Config.string(),
+					name: Config.string(),
+					options: Config.arrayOf(
+						Config.shapeOf(
+							{
+								label: Config.string(),
+								name: Config.string(),
+								value: Config.string()
+							}
+						)
+					),
+					type: Config.string(),
+					value: Config.string()
+				}
+			)
+		).value([]),
+
+		strings: {
+			value: {
+				addField: Liferay.Language.get('add-field')
+			}
+		}
+	}
+
+	attached() {
+		this._resetExpression();
+	}
+
+	willReceiveState(changes) {
+		let {expression, expressionArray} = this;
+
+		if (changes.hasOwnProperty('resultSelected')) {
+			this._resetExpression();
+		}
+		else if (changes.hasOwnProperty('options')) {
+			if (changes.expression) {
+				expression = changes.expression.prevVal;
+			}
+			else {
+				expression = '';
+			}
+
+			expressionArray = changes.expressionArray.prevVal;
+
+			this.setState(
+				{
+					expression,
+					expressionArray
+				}
+			);
+		}
+	}
+
+	_resetExpression() {
+		let {expression, expressionArray} = this;
+
+		expression = '';
+		expressionArray = [];
+
+		this.setState(
+			{
+				expression,
+				expressionArray
+			}
+		);
+	}
+
+	_formatExpression(expressionArray, expression) {
+		const regexRepeatableArithSigns = /[\+\-\*/]{2,}/;
+		const regexRepeatableDotSign = /[\.]{2,}/;
+
+		if (expression.match(regexRepeatableArithSigns) || expression.match(regexRepeatableDotSign)) {
+			const lastChar = expressionArray.pop();
+
+			expressionArray.pop();
+			expressionArray.push(lastChar);
+		}
+
+		return expressionArray;
+	}
+
+	_setFieldsRepeatable() {
+		const {options} = this;
+
+		let {optionsRepeatable} = this;
+
+		optionsRepeatable = options.filter(
+			option => {
+				return option.repeatable == true;
+			}
+		);
+
+		this.setState(
+			{
+				optionsRepeatable
+			}
+		);
+	}
+
+	_handleItemSelection(event) {
+		let {disableCalculatorField, expressionArray} = this;
+		let calculatorOperationFieldSelected = false;
```

Comment:
You don't need to declare `calculatorOperationFieldSelected  = false`, since undefined == false, you just need to declare `calculatorOperationFieldSelected = false` if you want to compare types, which is not the case. I saw that you only want to check if the `calculatorOperationFieldSelected` exists